### ISSUE TITLE
Move urpo DLL functionality to libcx

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/libcx#5]
It makes sense to move the functionality of the `urpo` DLL by @ydario (that implements unlinking and renaming open files in Linux style) into `libcx` because:
- both need to override the `close` function (see also #4).
- both need shared memory for inter-process operation and it's a waste of address space (and memory) to maintain two separate shared heaps (see also #3).

**Comment by dmik** [comment]
Also, the current `urpo` implementation lacks the `remove` override (which is equivalent to `unlink`, according to e.g. http://pubs.opengroup.org/onlinepubs/009695399/functions/remove.html). We should add it also. This missing thing may be the reason why we see a lot of temp files left over after some unix tools (like `yum`) that might use `remove` instead of `unlink`.

**Comment by dmik** [comment]
I've just realized that LIBCx can't behave like URPO by default — there are may be cases when the application wants to receive real errors from `unlink` and `rename` (e.g. EACCESS) instead of creating a deferred operation and getting no error from these functions. In fact, all applications made with native file behavior in mind (e.g. Qt) don't need deferred delete and should not get it. However, if we simply leave URPO overrides as is within LIBCx then any application linking against LIBCx (basically all of them in the near future as it's meant as a LIBC extension) will get URPO overrides by default — this is not acceptable. Strictly speaking, the deferred delete functionality provided by URPO is only needed by default for those Unix apps that are barely ported and not changed to go the native route regarding.

I have a solution though. LIBCx will provide renamed functions, e.g. `unlink_deferred`, `rename_deferred` etc so no deferred functionality will be present by default. However, we will also provide an `URPO.DLL` forwarder (was planned anyway) that will forward its `unlink` to `unlink_deferred` etc. So in order to get the deferred deletion the app will have to link to URPO (just as it does now).

**Comment by dmik** [comment]
Another thing to mention is that deferred deletion is only a simulation of what the Unix world has. There any open file will be deleted immediately, it's not a deferred operation (because files there are referred to by their inode and the file name is only one of many possible aliases for the inode that actually stores the file data — therefore the delete operation only deletes the alias, not the file itself). So even when URPO is used, incorrect behavior is still possible (one case is when the open file is deleted/renamed and then a new file is about to be created under the same name — URPO doesn't handle file opening so the open operation will simply fail). This is not related to this issue directly but should be considered at least.

**Comment by dmik** [comment]
In addition to above. Another inaccurate case is where deletion fails not due to the fact that the file is currently open but because the calling process simply does not have enough rights to delete the file (i.e the operation would fail even if the file were closed). This case is not handled at all and deletion will be deferred on such a file too hiding the failure from the application at all while the file won't be actually ever deleted (even after it is closed). Another failing case is when the file is open by another, non-LIBC process in which case deferred deletion will also not work (unless this another process terminates before the LIBC one).

All these failures make me think that URPO, at least in its current implementation, should be NOT used at all — instead, the application needs to be fixed to work the OS/2 way. We have to provide backward compatibility though...

**Comment by dmik** [comment]
I also have some problems with processing pending operations in the context of LIBCx. LBCx is made so that everything is uninitialized in either a normal call DLL_InitTerm when everything goes fine or in the process exit list handler when something goes wrong (a crash or so). However, URPO can't perform its pending operations in DLL_InitTerm as open files are still open by that time, it always needs the process exit list handler which is called very late after LIBC has uninitialized and even process files have been closed. Therefore I need to redo LIBCx to always do uninitialization from the process exit list handler rather than from DLL_InitTerm...

I really hope that LIBC heap manipulation code is still available after LIBC uninitialization.

**Comment by dmik** [comment]
Unfortunately, this has nothing to do with LIBC. Open file handles are closed by OS/2 at process termination far beyond LIBC control. In other words, there seems to be no way to execute any user code after OS/2 has forcefully closed open file handles. This means that the major part of URPO functionality (deleting files with unlink without calling close on them) can't be implemented at all w/o a dedicated daemon process that would process pending unlink regularly in hope the user application has already ended so that OS/2 closed all open file handles.

However, creating such a daemon is not a trivial task at all. Imagine all these possible multiple unlink/create calls on the same file from different processes and race conditions they could cause (just imagine a very simple one: process A unlinks file F and terminates, then process B creates file F and closes it; if the daemon kicks in between of these two processes it's all fine, if it gets its cycle after process B created and closed F, it will simply delete F which will be totally unexpected). I even think that creating a daemon that would function correctly is not possible at all given the means we have.

For this reason I voted for depreciation of the URPO library completely. The subsequent discussion with @SilvanScherrer and @ydario showed agreement on this topic. Hence, using of URPO is now officially deprecated and it makes no sense to integrate it with LIBCx. It means that new app releases may not link against URPO from now on and, especially, URPO must never be used with LIBCX (as it will break some of its overrides, e.g. `close`, and related functionality).

For historical reasons, I committed my current URPO integration in 5478624b70fde2b20ca05ecfd4e13346eca5b13e but in the disabled state. I will remove this integration (together with URPO code) completely in a subsequent commit which will also close this task.
